### PR TITLE
ARM Mali GPU Panfrost support

### PR DESCRIPTION
**Pull request by larunbe**
Add support for ARM Mali GPUs monitoring with the Panfrost driver.

### CODE REVIEW

**Comment by larunbe** [comment]
> Hi, I tried to build but failed due to https://github.com/larunbe/nvtop/blob/bb580d3801c6a0f9f3807740d9098d4df4450b76/src/extract_gpuinfo_panfrost_utils.c#L22. File `panfrost_utils.h` seems do not exist. I removed that line and build passed. Is it a bug?

Sorry, I recently added that file but forgot to include it in the PR update. It's been included in the latest push.

**Review by Syllo — COMMENTED** [review]
Thanks for this patch. Looks good to me so far.
There is only one thing missing in my opinion is a toggle in the F2 setup menu (see [header options](https://github.com/Syllo/nvtop/blob/487858b1120f4a3ed05e14ca4b8944f5964251d9/src/interface_options.c#L213)). I'll add it this week if I have some time to spare.

**Comment by larunbe** [comment]
> Thanks for this patch. Looks good to me so far. There is only one thing missing in my opinion is a toggle in the F2 setup menu (see [header options](https://github.com/Syllo/nvtop/blob/487858b1120f4a3ed05e14ca4b8944f5964251d9/src/interface_options.c#L213)). I'll add it this week if I have some time to spare.

Is this F2 functionality present in other drivers? If so, I could do it myself.

File: include/nvtop/extract_gpuinfo_common.h [hunk 1]
```diff
@@ -66,6 +66,9 @@ struct gpuinfo_static_info {
   unsigned max_pcie_link_width;
   unsigned temperature_shutdown_threshold;
   unsigned temperature_slowdown_threshold;
+  unsigned n_shared_cores;
```
[the remark this answers is not in the document]
Comment (larunbe) [reply include/nvtop/extract_gpuinfo_common.h:69]: I guess you'd also want me to set these values with `SET_GPUINFO_STATIC` ?

[the remark this answers is not in the document]
Comment (larunbe) [reply include/nvtop/extract_gpuinfo_common.h:72]: This is fixed in the latest PR update.

**Comment by nextfoe** [comment]
Hi @larunbe ,
May I run nvtop on the embedded devices or android phone after cherry-pick your patches?
Now i has a issue:
```
adb shell
# ./data/nvtop  -i
profile parameter not implemented in Panfrost
No GPU to monitor.
```

Thanks

**Comment by larunbe** [comment]
> Hi @larunbe , May I run nvtop on the embedded devices or android phone after cherry-pick your patches? Now i has a issue:
> 
> ```
> adb shell
> # ./data/nvtop
> profile parameter not implemented in Panfrost
> No GPU to monitor.
> 
> # ls /sys/kernel/debug/dri/128/
> no profile file, just has a name file.
> ```
> 
> Thanks

You're going to need `drm-misc` to run this on an SoC with a Mali GPU.

**Comment by nextfoe** [comment]
> > Hi @larunbe , May I run nvtop on the embedded devices or android phone after cherry-pick your patches? Now i has a issue:
> > ```
> > adb shell
> > # ./data/nvtop
> > profile parameter not implemented in Panfrost
> > No GPU to monitor.
> > 
> > # ls /sys/kernel/debug/dri/128/
> > no profile file, just has a name file.
> > ```
> > 
> > 
> > Thanks
> 
> You'll need the latest changes on `drm-misc` to run this on an SoC with a Mali GPU. The patch submission is here: https://lore.kernel.org/lkml/20231004130601.60239c7f@collabora.com/ However like I said the patch series has already landed in the drm-misc tree, so I guess that means it'll be part of the next 6.7 release.

Thanks, I will try cherry-pick this patch to our kernel version to test again.

File: include/nvtop/extract_gpuinfo_common.h [hunk 2]
```diff
@@ -184,9 +197,11 @@ struct gpu_vendor {
 
   void (*populate_static_info)(struct gpu_info *gpu_info);
   void (*refresh_dynamic_info)(struct gpu_info *gpu_info);
+  void (*refresh_utilisation_rate)(struct gpu_info *gpu_info);
 
   void (*refresh_running_processes)(struct gpu_info *gpu_info);
   char *name;
+  /* void *priv; */
```
Comment (Syllo) [inline include/nvtop/extract_gpuinfo_common.h:204]: Do you plan on using this or can the comment be removed?

File: src/extract_gpuinfo_mali_common.c [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+      return;
+    cache_entry->client_id.client_id = cid;
+    cache_entry->client_id.pid = process_info->pid;
+    cache_entry->last_cycles = total_cycles;
+  }
+
+#ifndef NDEBUG
+  // We should only process one fdinfo entry per client id per update
+  struct mali_process_info_cache *cache_entry_check;
+  HASH_FIND_CLIENT(gpu_info->current_update_process_cache, &cid, cache_entry_check);
```
Comment (Syllo) [inline src/extract_gpuinfo_mali_common.c:427]: We fixed some uniqueness issue for the id in #248, I'll push a patch to the pull request to apply the same fix.

File: src/panthor_drm.h [hunk 4]
```diff
@@ -0,0 +1,876 @@
+/* SPDX-License-Identifier: MIT */
```
Comment (Syllo) [inline src/panthor_drm.h:1]: Shouldn't this header part of the Linux headers?
I guess that if it's too recent we can keep the header until it's widely available if it's stable enough

**Comment by Syllo** [comment]
I pushed a patch to fix the cache lookup and two others to get the "Extra GPU info line" in the F2 menu and made it so that it gets saved/loaded from the config file.

File: include/nvtop/extract_gpuinfo_common.h (hunk 2 above)
Comment (larunbe) [reply include/nvtop/extract_gpuinfo_common.h:204]: Sorry about this, it was going to implement some sort of backend-specific private data flow and forgot to delete it before the PR.

File: src/panthor_drm.h (hunk 4 above)
Comment (larunbe) [reply src/panthor_drm.h:1]: My understanding is that this sort of uAPI file isn't part of the usual kernel headers package under /usr/include/linux. Both Mes and the IGT test suit have to keep the same uAPI files up to date with the one in the kernel sources, so I thought we'd have to do the same for nvtop.

**Comment by larunbe** [comment]
If you want to try the Panthor backend, you'd need a SBC with a Mali CSF GPU (Like RockPi5) and the fdinfo enablement branch I'm currently working on: https://gitlab.freedesktop.org/larumbe/panthor/-/tree/panthor-fdinfo

**Comment by larunbe** [comment]
Hi @Syllo, do you think you might have a wee bit of time to have a peek into my latest changes some time this week?
Cheers.

**Comment by Syllo** [comment]
Hey @larunbe,

I think that it's in a good enough state to be included. Do you wish to add something more before merging?

**Comment by larunbe** [comment]
Hi @Syllo,
Someone running nvtop on a SBC with a Mali CSF GPU has noticed when unsetting `XDG_CONFIG_HOME` it crashes in a segmentation fault.
I've been unable to reproduce it on my local setup. They'll be giving me a full stacktrace of the event next week, and when I've figured out what's wrong and fixed it, will let you know so that you can merge the changes.
Cheers

**Pull request merged**